### PR TITLE
feat: improve `Trainer` and `DeeprankDataset` logic for production testing

### PR DESCRIPTION
**Pull request by gcroci2**
**Main changes:**
- Now `DeeprankDataset` takes as input `train_data`, that before was called `dataset_train`. Now `train_data` can be a `DeeprankDataset` representing the training set (as before), or a pre-trained model (new feature). It needs to be set only if `train` is False (as before, so only in validation/testing sets cases). Now we are able to use a test dataset without the need for the original model's training dataset. We can use the info stored in the pre-trained model to inherit the needed attributes.
  - There were many issues with loading the lambda transformations from the torch-saved pre-trained model (testing only use case). I ended up converting the lambdas to strings first, and then saving them. When the pre-trained model is loaded in the `GraphDataset` class, the strings representing the lambdas are evaluated and converted back to functions. 

**Secondary changes:**
- In `DeeprankDataset` classes, if `target` attribute is present (e.g., `binary`, inherited) but it's not in the HDF5 _and_ we're not in the training phase, now no error is raised. Indeed, it should be possible to run a pre-trained model on data point/s even if the target value/s are not present, for doing predictions only. It's actually a typical test-case scenario, in which we don't have any labels for the new data points that we want to evaluate. 
- On the other hand, if we're in the training phase (`train = True`) and no target is set, or the set target is not in the hdf5 file/s, then a `ValueError` is raised. 
- `self.pretrained_model_path` is now defined in the init of the `Trainer` class and defaulted to `pretrained_model`.
- `self.model_load_state_dict` is also defined in the init of the `Trainer` class and defaulted to `None`. It is assigned to a value only in the case of a pre-trained model or at the end of the training phase. This way in the `test()` method we can first verify if the model has actually been loaded (pre-trained case) or trained. If not, the `test()` method now throws an error. 
- In the Trainer class' init, before loading parameters and the pre-trained model there was a check for the target, which in the pre-trained model case could be not present at all (it is saved in the model itself, no need to define it in the `Trainer` instance). I removed the check.
- I added the following attributes in `_init_from_dataset`, which need to be saved in the model's file for those cases in which we want to test it on some other data without redefining the training set: `features_transform`, `means`, `devs`, `target_transform`, `classes`, `classes_to_index`. I also added them to the model's `state` dict which is saved at the end of the training (`_save_model`, same for `_load_params`).
- Now the saved model contains a key called `data_type`, needed for checking which type of dataset was used during the training of the model. 
- I removed the warning about not having a validation set during the training because it was given at _each_ epoch. Now it's printed only once when you call the `train()` method.
- When you call torch.load() is called on a model's file which contains GPU tensors, those tensors will be loaded to GPU by default. But if no Cuda was available, the code crashed. Now there is a check for that, and in case Cuda is not available the tensors are loaded into the CPU.
- `target_filter` wasn't really working. Some other edits made me notice that the functionality was broken, and it's fixed now. 

Still to solve:
- For some reason, PyTorch gives a weird error about the Adam optimizer, but only in the 3.11 Python version. I tried to fix the relevant torch packages versions, but it's still failing. I haven't touched anything about the optimizer though. Any idea? @DaniBodor

### CODE REVIEW

File: deeprank2/trainer.py [hunk 1]
```diff
@@ -210,30 +212,33 @@ def _init_datasets(self,  # pylint: disable=too-many-arguments
...
         elif isinstance(dataset, GridDataset):
             self.clustering_method = None
             self.node_features = None
             self.edge_features = None
             self.features = dataset.features
+            self.features_transform = None
+            self.means = None
+            self.devs = None
         else:
             raise TypeError(type(dataset))
```
Comment (DaniBodor) [inline deeprank2/trainer.py:235]: Would be nice to return a better error message here. Maybe something like (although not perfect, because to the user it's unclear what `dataset` is in this case).:

```suggestion
        else:
    raise TypeError(
        f"Incorrect `dataset` type provided: {type(dataset)}. Please provide a `GridDataset` or `GraphDataset` object instead."
    )
```

File: deeprank2/trainer.py [hunk 2]
```diff
@@ -739,6 +748,9 @@ def _eval( # pylint: disable=too-many-locals
                 loss_ = loss_func(pred, y)
                 count_predictions += pred.shape[0]
                 sum_of_losses += loss_.detach().item() * pred.shape[0]
+            else:
+                target_vals += ['None'] * pred.shape[0]
+                eval_loss = 'None'
```
Comment (DaniBodor) [inline deeprank2/trainer.py:753]: Why are these strings instead of a `None`? Same question below.

I also noticed that the `_eval` method returns this `eval_loss`, but that 3/4 times the method is called, it looks like the return value is not stored. 
I am not sure if the `OutputExporterCollection` is dealing with that as I am not perfectly clear on what the `with self._output_exporters` line does. If it does, I think it would be useful to document that in the code somehow, either in the docstring of `_eval` or as a comment in the code.

Finally, the typehint for the return value should be updated in the definition of the code to allow either type.

File: tests/test_dataset.py
[the remark this answers is not in the document]
Comment (DaniBodor) [reply tests/test_dataset.py:1]: Same applies to test_trainer.py as well.

**Review by DaniBodor — COMMENTED** [review]
Great additions. It's really nice that users can now just run an analysis on their data without doing a whole training session. I think this will end up being the main way people will interact with the software, and that a full training will be more like the "superuser" way of interacting with it.
I would change the type of this PR to feat. It's not really a refactoring (at least not primarily), but implementing new functionality.

To me, the most important comment to address is this one [placeholder]. Maybe we can think together of an elegant way to do this (or maybe you have a good idea already.

I think the modules and the tests could do with a refactoring makeover, but that is not high priority at this point and not sure we will get to it at all anymore and def don't have to be addressed in this PR.

I haven't looked at all yet at the list of things left open. Let me know when you want to discuss those.

File: docs/getstarted.md [hunk 3]
```diff
@@ -412,3 +412,68 @@ fig.update_layout(
     title='Loss vs epochs'
 )
 ```
+
+## Testing new data
+
+If new PDB files need to be tested with a pre-trained model, the first step would be to process and save them into HDF5 files. Let's suppose that the model has been trained with `ProteinProteinInterfaceResidueQuery` queries mapped to graphs:
+
+```python
+from deeprank2.query import QueryCollection, ProteinProteinInterfaceResidueQuery
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply docs/getstarted.md:420]: Indeed, I just merged with dev

File: deeprank2/dataset.py [hunk 4]
```diff
@@ -54,11 +56,11 @@ def __init__(self, # pylint: disable=too-many-arguments
...
 
-        self.use_tqdm = use_tqdm
-
+        self.train = train
+        self.train_data = train_data
         self.target = target
         self.subset = subset
-
+        self.use_tqdm = use_tqdm
         self.target_filter = target_filter
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py]: Thanks, nice catch :) I fixed it in this PR (was not a lot); now we have `use_tqdm` and `root_directory_path` only, and also the order should be now coherent across the classes in dataset.py.

File: deeprank2/dataset.py [hunk 5]
```diff
@@ -429,28 +434,28 @@ def __init__( # pylint: disable=too-many-arguments
...
-                The latter will be used to scale the validation/testing set according to its features values and to match the datasets' parameters.
+                It needs to be set to False only if the instance represents the validation or testing set.
                 Defaults to True.
-            dataset_train (class:`GridDataset`, optional): If `train` is True, assign here the training set.
-                If `train` is False and `dataset_train` is assigned,
-                the parameters `features`, `target`, `traget_transform`, `task`, and `classes` will be inherited from `dataset_train`.
+            train_data (Optional[Union[str, class:`GridDataset`]], optional)): data to inherit information from the training dataset or the pre-trained model.
+                It is considered only if `train` is False. In the latter case, `train_data` needs to be a dataset of the same class or
+                the path of a DeepRank2 pre-trained model. If set, the parameters `features`, `target`, `traget_transform`, `task`, and `classes`
+                will be inherited from `train_data`.
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py:444]: I like the idea of using only one parameter (maybe we can think of a better naming?) that can be either None, a pre-trained model, or a dataset_train. It seems much easier to me also code-wise, and less redundant. Let me know what you think about this option, and if you have an idea for better naming (maybe `train_source`? Or maybe `train_data` is good enough)

File: deeprank2/dataset.py [hunk 6]
```diff
@@ -698,45 +745,82 @@ def __init__( # noqa: MC0001, pylint: disable=too-many-arguments, too-many-local
...
+                except pickle.UnpicklingError as e:
+                    raise ValueError("""The path provided to `train_data` is not a valid DeepRank2 pre-trained model.
+                                            Please provide a valid path to a DeepRank2 pre-trained model.""") from e
+            elif isinstance(train_data, GraphDataset):
+                train_means = train_data.means
+                train_devs = train_data.devs
+                data = train_data
+            else:
+                raise TypeError(f"""The train data provided is type: {type(train_data)}
+                                Please provide a valid training GraphDataset or the path to a valid DeepRank2 pre-trained model.""")
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py]: if `data` does not exist, then we can't assign `train_means` and `train_devs`. 

I wrapped all the code for doing this part into the parent class, in `_check_and_inherit_train` method. Nicer indeed :)

File: deeprank2/dataset.py [hunk 7]
```diff
@@ -595,22 +627,36 @@ def load_one_grid(self, hdf5_path: str, entry_name: str) -> Data:
...
 
         feature_data = []
-        target_value = None
 
         with h5py.File(hdf5_path, 'r') as hdf5_file:
-            entry_group = hdf5_file[entry_name]
+            grp = hdf5_file[entry_name]
 
-            mapped_features_group = entry_group[gridstorage.MAPPED_FEATURES]
+            mapped_features_group = grp[gridstorage.MAPPED_FEATURES]
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py:654]: `grp` stays for Group, which is the container mechanism by which HDF5 files are organized. In our case, each group is an entry that represents a graph (a single data point/complex). Also in the official `h5py` [documentation](https://docs.h5py.org/en/stable/high/group.html), grp is widely used to indicate single Groups in the HDF5 files. 

I uniformed `entry_group` with `grp` all across the code-base now.

File: tests/test_dataset.py [hunk 8]
```diff
@@ -220,18 +221,18 @@ def test_inherit_info_training_griddataset(self):
...
             train = False,
-            dataset_train = dataset_train
+            train_data = dataset_train
         )
 
-        # features, features_dict, target, target_transform, task, and classes
+        # features, target, target_transform, task, and classes
         # in the test should be inherited from the train
-        inherited_param = ["features", "features_dict", "target", "target_transform", "task", "classes"]
+        inherited_param = ["features", "target", "target_transform", "task", "classes"]
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply tests/test_dataset.py:229]: now `inherited_param` is an attribute of the datasets

File: tests/test_query.py [hunk 9]
```diff
@@ -39,6 +39,7 @@ def _check_graph_makes_sense(g, node_feature_names, edge_feature_names):
     os.close(f)
 
     try:
+        g.targets[targets.BINARY] = 0
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply tests/test_query.py:46]: because `_check_graph_makes_sense` defines a GraphDataset instance that is in training mode, and now this requires a target to be defined (otherwise you can't train a model)

**Comment by DaniBodor** [comment]
I unsubscribed to notifications for this PR for now. Please tag me again if needed and/or when you want me to re-review.

File: tests/test_dataset.py [hunk 10]
```diff
@@ -243,15 +244,111 @@ def test_inherit_info_training_griddataset(self):
...
+            target_transform = True,
+            task = targets.REGRESS,
+            classes = None
+        )
+
+        # features, target, target_transform, task, and classes
+        # in the test should be inherited from the pre-trained model
+        dataset_test_vars = vars(dataset_test)
+        for param in inherited_params:
+            assert dataset_test_vars[param] == data[param]
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply tests/test_dataset.py]: The first part is to test the inheritance not giving in any parameters. The second part that you highlighted is for checking that even if you give in different parameters from the training data, the inheritance works.

File: tests/test_dataset.py
[the remark this answers is not in the document]
Comment (gcroci2) [reply tests/test_dataset.py:1]: Opened an issue about that https://github.com/DeepRank/deeprank2/issues/526

File: docs/getstarted.md [hunk 11]
```diff
@@ -412,3 +412,68 @@ fig.update_layout(
     title='Loss vs epochs'
 )
 ```
+
+## Testing new data
+
+If new PDB files need to be tested with a pre-trained model, the first step would be to process and save them into HDF5 files. Let's suppose that the model has been trained with `ProteinProteinInterfaceResidueQuery` queries mapped to graphs:
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply docs/getstarted.md:417]: In the ML/DL terminology, "testing" usually refers to data that are not used during the training. New data can be considered new testing cases. I changed the title to "## Run a pre-trained model on new data" which I think is clearer.

[the remark this answers is not in the document]
Comment (gcroci2) [reply docs/getstarted.md:417]: Regarding the repetition, I think it's better to repeat the code block. A user may skip the rest and may be confused. Also, it's not in the readme, so it's fine if it's longer

File: docs/getstarted.md [hunk 12]
```diff
@@ -412,3 +412,68 @@ fig.update_layout(
...
+    chain_id1 = "A",
+    chain_id2 = "B"
+))
+
+hdf5_paths = queries.process(
+    "<output_folder>/<prefix_for_outputs>",
+    feature_modules = 'all')
+```
+
+Then, the GraphDataset instance representing the testing set can be defined. Note that there is no need of setting the dataset's parameters, since they are inherited from the information saved in the pre-trained model.
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply docs/getstarted.md:441]: As above, new data points are new testing points, so in ML/DL gergon the dataset can be considered a new "testing" dataset (for sure not a training set). It only _inherits_ parameters from the training data, but it contains testing data. Also in this case I did a mix between the old version and your suggestions.

File: docs/getstarted.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (gcroci2) [reply docs/getstarted.md]: Added to the readme. I think it will be nice to add a tutorial only when we publish a pre-trained model (such as the one used in the 3D-Vac paper), or when we create a more "advanced" tutorial. For now having that in the .md files only is fine.

File: docs/getstarted.md [hunk 13]
```diff
@@ -411,3 +411,68 @@ fig.update_layout(
...
+
+queries = QueryCollection()
+
+# Append data points
+queries.add(ProteinProteinInterfaceResidueQuery(
+    pdb_path = "<new_pdb_file1.pdb>",
+    chain_id1 = "A",
+    chain_id2 = "B"
+))
+queries.add(ProteinProteinInterfaceResidueQuery(
```
Comment (DaniBodor) [inline docs/getstarted.md:430]: As this part of the docs is new in this PR, the class names didn't get updated when you merged with dev.

```suggestion
If you want to run a pre-trained model on new PDB files, the first step is to process and save them into HDF5 files. Let's suppose that the model has been trained with `ProteinProteinInterfaceQuery` queries mapped to graphs:

```python
from deeprank2.query import QueryCollection, ProteinProteinInterfaceQuery

queries = QueryCollection()

# Append data points
queries.add(ProteinProteinInterfaceQuery(
    pdb_path = "<new_pdb_file1.pdb>",
    chain_id1 = "A",
    chain_id2 = "B"
))
queries.add(ProteinProteinInterfaceQuery(
```

File: deeprank2/dataset.py [hunk 14]
```diff
@@ -595,22 +627,36 @@ def load_one_grid(self, hdf5_path: str, entry_name: str) -> Data:
...
+                if targets.VALUES in grp and self.target in grp[targets.VALUES]:
+                    y = torch.tensor([grp[targets.VALUES][self.target][()]], dtype=torch.float)
 
-        # Wrap up the data in this object, for the collate_fn to handle it properly:
-        data = Data(x=torch.tensor(np.expand_dims(np.array(feature_data), axis=0), dtype=torch.float),
-                    y=torch.tensor([target_value], dtype=torch.float))
+                    if self.task == targets.REGRESS and self.target_transform is True:
+                        y = torch.sigmoid(torch.log(y))
+                    elif self.task is not targets.REGRESS and self.target_transform is True:
+                        raise ValueError(f"Task is set to {self.task}. Please set it to regress to transform the target with a sigmoid.")
```
[the remark this answers is not in the document]
Comment (DaniBodor) [reply deeprank2/dataset.py]: My question here was whether we should also make the argument name something more specific than `target_transform`. What do you think?
Maybe something along the lines of `normalize_target` or `target_normalization`? It is still somewhat generic, but I think clearer and more informative about what the use of this argument is.

File: tests/test_dataset.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply tests/test_dataset.py]: I think it would be useful to put a short comment above it to explain this.

File: docs/getstarted.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply docs/getstarted.md]: agreed

**Review by DaniBodor — COMMENTED** [review]
Just leaving these as comments for now. Once you/we figure out why the build is failing, I will review that before approving.

File: deeprank2/dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py]: @DaniBodor

File: deeprank2/dataset.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gcroci2) [reply deeprank2/dataset.py]: I think is good enough for now (we're actually never using it), and it is coherent with the features transform nomenclature

File: deeprank2/trainer.py (hunk 2 above)
Comment (gcroci2) [reply deeprank2/trainer.py]: Not storing the return value doesn't give particular issues, and it exists in the local scope of the methods only

File: tests/test_dataset.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (gcroci2) [reply tests/test_dataset.py]: I'll do

File: deeprank2/dataset.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply deeprank2/dataset.py]: We agreed to remove the current `train` argument, as it is redundant with the current `train_data` argument. If `train_data is None`, that means that means the same as `train is True`. Otherwise it means `train is False`.

Also, we agreed to rename it to `train_source` until we come up with a better name for the parameter.

**Comment by DaniBodor** [comment]
It looks to me like the problem with the 3.11 build is really a core change in pytorch. I don't think it'll be easy for us (def not me) to figure out what the problem is. Maybe it's best to create an issue on pytorch and see if they know how to solve.

**Comment by gcroci2** [comment]
I am merging this PR. The issue with Python 3.11 will be solved in another PR.
